hnix-store-core/tests/Hash.hs

```diff
@@ -41,16 +41,16 @@ spec_hash = do
       let exampleStr =
             "source:sha256:2bfef67de873c54551d884fdab3055d84d573e654efa79db3"
             <> "c0d7b98883f9ee3:/nix/store:myfile"
-      shouldBe (encodeBase32 @StorePathHashAlgo (hash exampleStr))
+      shouldBe ((encodeIn Base32) @StorePathHashAlgo (hash exampleStr))
```

Comment:
Interesting encounter:

Old code:
```haskell
      shouldBe (encodeBase32 @StorePathHashAlgo (hash exampleStr))
        "xv2iccirbrvklck36f1g7vldn5v58vck"
       -- where
       --  encodeBase32 :: Digest a -> T.Text
```

New code:
```haskell
      shouldBe ((encodeIn Base32) @StorePathHashAlgo (hash exampleStr))
        "xv2iccirbrvklck36f1g7vldn5v58vck"
       -- where
       --  encodeIn :: BaseEncoding -> Digest a -> T.Text
       -- so
       --  encodeIn Base32 :: Digest a -> T.Text
```

Recieves:
```
tests/Hash.hs:44:17: error:
    • Cannot apply expression of type ‘Digest a0
                                       -> Data.Text.Internal.Text’
      to a visible type argument ‘StorePathHashAlgo’
    • In the first argument of ‘shouldBe’, namely
        ‘((encodeIn Base32) @StorePathHashAlgo (hash exampleStr))’
      In a stmt of a 'do' block:
        shouldBe
          ((encodeIn Base32) @StorePathHashAlgo (hash exampleStr))
          "xv2iccirbrvklck36f1g7vldn5v58vck"
      In the second argument of ‘($)’, namely
        ‘do let exampleStr
                  = "source:sha256:2bfef67de873c54551d884fdab3055d84d573e654efa79db3"
                      <> "c0d7b98883f9ee3:/nix/store:myfile"
            shouldBe
              ((encodeIn Base32) @StorePathHashAlgo (hash exampleStr))
              "xv2iccirbrvklck36f1g7vldn5v58vck"’
   |
44 |       shouldBe ((encodeIn Base32) @StorePathHashAlgo (hash exampleStr))
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

I need to lookup, more what it is about, on the face seems really strange.

Reply:
Type application `@StorePathHashAlgo` has to be the first "argument".

Your type is `encodeIn :: BaseEncoding -> Digest a -> T.Text` which due to `ScopedTypeVariables` (which implies `ExplicitForAll`) gets rewritten to `encodeIn :: forall a . BaseEncoding -> Digest a -> T.Text` and using `TypeApplications` you need to apply algorithm to `a` type var so it's not ambiguous.

Reply:
Probably for you it is something obvious, but for me - it is completely obscure and needs reading.

Since the promise of functions being the first class, the equivalent function sectioning should be equivalent. Thanks for the pointer into the direction to understand.
I definitely need to look deeper into quantifiers and types.

Currently the easiest/prettiest solution - is to `where` the section application :rofl:

Reply:
Ok.

I got the interplay between the language extensions and quantifiers.

I forgot what `ScopedTypeVariables` does, thou the name is instructive when one remembers.

The caveat seems to be that visible type application happens to an invisible type argument that GHC exposes to a declared or annotated function, otherwise, GHC seems to not give access to the invisible type application argument before argument.

Reply:
Cool, I've learned about this while working on this codebase and toying with polysemy and my toy polytype lib on top of it. It takes some time to get used to type level programming in Haskell :-)

Reply:
Type level programming in Haskell:

Me: So how many extensions do you want?
GHC: Yes.

Reply:
TypeApplicaton with DataKinds is a strange beast.

I want my kinds to be `* -> *` and even better - sometimes polymorphic, so `*` accepts everything. I want my kinds to be `* ` and `* -> *` not `HashAlgorithm` suddenly. What is a `Hash Algorithm` kind, not calling the harder cases. When types suddenly become not types and kinds suddenly become not kinds - how I should think about them.

DataKinds is an extension from the dependent types lambda space, btw "Giving Haskell a Promotion" has an argumentation against dragging-in all functionality from dependent types.

I am OK and cool with native lambda field extensions.
And as I see, dependent type extensions need to be used one, to several at a time in an enclosed environment. If someone somewhere wants to go full dependent types - do that in self-enclosed environment so the interface is classic Haskell (typed parametric lambda calculus). Because DT extensions may easily slip across the project - and suddenly the code is harder and becomes rigid, not refactorable and refactoring becomes not evolution but a ripping-out of that system.

Reply:
Also many of the extensions are shown and discussed with a lab coat.
And nobody explained what that extension can be effective for in life application.

In what classical example TypeApplication with DataKinds really make code more performant and to what degree, I understand that it allowed to move some computation to compile time, but newcomer needs to know "how much" comparing to the cost of the introduced code rigidity and complexity.